ambry-router/src/test/java/com/github/ambry/router/CloudOperationTest.java

```diff
@@ -0,0 +1,881 @@
+    byte[] userMetadata = new byte[10];
+    random.nextBytes(userMetadata);
+    byte[] putContent = new byte[blobSize];
+    random.nextBytes(putContent);
+    BlobId blobId = doPut(blobProperties, userMetadata, putContent);
+
+    // The put for the blob being requested happened.
+    String dcWherePutHappened = routerConfig.routerDatacenterName;
+
+    // GetBlobOperation to DC1 returns Not Found. Then will get blob from the DC3 where put happened.
+    Properties props = getDefaultNonBlockingRouterProperties(true, "DC1");
+    routerConfig = new RouterConfig(new VerifiableProperties(props));
+    doTestSuccessInThePresenceOfVariousErrors(blobId, dcWherePutHappened, blobSize, blobProperties, userMetadata, putContent);
+
+    // DC2 returns different errors. Then will get blob from the DC3 where put happened.
+    props = getDefaultNonBlockingRouterProperties(true, "DC2");
+    routerConfig = new RouterConfig(new VerifiableProperties(props));
+    doTestSuccessInThePresenceOfVariousErrors(blobId, dcWherePutHappened, blobSize, blobProperties, userMetadata, putContent);
+
+    // test requests coming in from local dc.
+    props = getDefaultNonBlockingRouterProperties(true, "DC3");
+    routerConfig = new RouterConfig(new VerifiableProperties(props));
+    doTestSuccessInThePresenceOfVariousErrors(blobId, dcWherePutHappened, blobSize, blobProperties, userMetadata, putContent);
+  }
+
+  /**
+   * Disk backed DC all returns failure but cloud backed DC returns the Blob information successfully.
+   */
+  @Test
+  public void testGetBlobFailoverToAzure() throws Exception {
+    // a blob size that is greater than the maxChunkSize and is not a multiple of it. Will result in a composite blob.
+    int blobSize = maxChunkSize * random.nextInt(10) + random.nextInt(maxChunkSize - 1) + 1;
+    BlobProperties blobProperties = new BlobProperties(blobSize, "serviceId", "memberId", "contentType", false, Utils.Infinite_Time,
+       Utils.getRandomShort(random), Utils.getRandomShort(random), testEncryption, null, null, null);
+    byte[] userMetadata = new byte[10];
+    random.nextBytes(userMetadata);
+    byte[] putContent = new byte[blobSize];
+    random.nextBytes(putContent);
+    ByteBuf putContentBuf = PooledByteBufAllocator.DEFAULT.heapBuffer(blobSize);
+    putContentBuf.writeBytes(putContent);
+    BlobId blobId = doDirectPut(BlobType.DataBlob, blobProperties, userMetadata, putContentBuf.retainedDuplicate());
+    putContentBuf.release();
+
+    // Confirm we can get the blob from the local dc.
+    GetBlobOptionsInternal options = new GetBlobOptionsInternal(new GetBlobOptionsBuilder().build(), false, routerMetrics.ageAtGet);
+    getBlobAndAssertSuccess(blobId, false, false, (short)0, blobSize, blobProperties, userMetadata, putContent, options);
+
+    // Local DC will fail with different errors but cloud will return the blob data.
+    ArrayList<MockServer> mockServers = new ArrayList<>(mockServerLayout.getMockServers());
```

Comment:
`mockServers` would include not only the disk based server, but also the cloud servers. So in this case, even the cloud would return NOT_FOUND. 

`MockServerLayout` would construct a list of `MockServer`s in it's constructor. This element in the list matches to the `DataNode` from `MockClusterMap`, one to one. And `DataNode` in `MockClusterMap` actually includes the cloud data node.

Reply:
Sorry I made a mistake on the cloud data node. The request to cloud replica would be handled by the LocalNetworkClient, not the MockServers. 

That being said, if we want to make cloud replica return error, we can't rely on the mock server layout, we have to some how inject the error code to the local network client.

Reply:
Sorry. Here the Mock Server is a little bit confusing and it's something I have some concern and mentioned in the PR comment. I'll add more comment to the code.
For now, Ambry and cloud has different mock nodes, mock interface/implementation.
We do generate MockServers for each nodes. But for Ambry and Cloud, we have different network client to talk with the mock storage nodes.
We are using MockCompositeNetworkClient. This is a composite client with two clients, one is Ambry Mock SocketNetworkClient, the other is cloud LocalNetworkClient. Only Mock SocketNetworkClient is hooked with MockServer. LocalNetworkClient doesn't. So the simulated server status won't affect cloud. For cloud, it will only return NOT_FOUND or success depending on if we put blob in cloud.

Following is the call stack for two cases,
#######################For Ambry,
  java.lang.Thread.State: RUNNABLE
	  at com.github.ambry.router.MockServer.makeGetResponse(MockServer.java:184)
	  at com.github.ambry.router.MockServer.send(MockServer.java:106)
	  at com.github.ambry.router.MockSelector.poll(MockSelector.java:123)
	  at com.github.ambry.network.SocketNetworkClient.sendAndPoll(SocketNetworkClient.java:127)
	  at com.github.ambry.network.CompositeNetworkClient.lambda$sendAndPoll$1(CompositeNetworkClient.java:97)
	  at com.github.ambry.network.CompositeNetworkClient$$Lambda$88.2061832550.call(Unknown Source:-1)
	  at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	  at java.util.concurrent.FutureTask.run(FutureTask.java:-1)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	  at java.lang.Thread.run(Thread.java:834)


#######################For Cloud,
  java.lang.Thread.State: RUNNABLE
	  at com.github.ambry.network.LocalRequestResponseChannel.byteBufFromPayload(LocalRequestResponseChannel.java:127)
	  at com.github.ambry.network.LocalNetworkClient.sendAndPoll(LocalNetworkClient.java:83)
	  at com.github.ambry.network.CompositeNetworkClient.lambda$sendAndPoll$1(CompositeNetworkClient.java:97)
	  at com.github.ambry.network.CompositeNetworkClient$$Lambda$88.2061832550.call(Unknown Source:-1)
	  at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	  at java.util.concurrent.FutureTask.run(FutureTask.java:-1)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	  at java.lang.Thread.run(Thread.java:834)

Reply:
Agree. We should have some error injection for LocalNetworkClient code path as well.

Reply:
Supported cloud mock layer error injection in another PR. 
Will rebase upon on it when the "error simulation PR" got approved.